cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -427,6 +427,189 @@ void AddMatchSubstring(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::move(func)));
 }
 
+// Slicing
+
+template <typename Type, typename Derived>
+struct SliceBase : StringTransform<Type, Derived> {
+  using Base = StringTransform<Type, Derived>;
+  using offset_type = typename Base::offset_type;
+  using State = OptionsWrapper<SliceOptions>;
+  SliceOptions options;
+
+  explicit SliceBase(SliceOptions options) : options(options) {}
+
+  static void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    SliceOptions options = State::Get(ctx);
+    if (options.step == 0) {
+      ctx->SetStatus(Status::Invalid("Slice step cannot be zero"));
+      return;
+    }
+    Derived(options).Execute(ctx, batch, out);
+  }
+
+  void Execute(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    Base::Execute(ctx, batch, out);
+  }
+};
+
+#define PROPAGATE_FALSE(expr)         \
+  do {                                \
+    if (ARROW_PREDICT_FALSE(!expr)) { \
+      return false;                   \
+    }                                 \
+  } while (0)
+
+template <typename Type>
+struct SliceCodeunits : SliceBase<Type, SliceCodeunits<Type>> {
+  using Base = SliceBase<Type, SliceCodeunits<Type>>;
+  using offset_type = typename Base::offset_type;
+  using Base::Base;
+
+  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
+                 uint8_t* output, offset_type* output_written) {
```

Comment:
Note that most of this function is really identical regardless of the offset type. So instead you could define a `Utf8SliceBase` base class:
```c++
struct Utf8SliceBase {
  bool Transform(const uint8_t* input, int64_t input_string_ncodeunits,
                 uint8_t* output, int64_t* output_written) {
    // ...
  }
};
```
... and inherit from that mixin class in concrete classes:
```c++
template <typename Type>
struct SliceCodeunits : SliceBase<Type, SliceCodeunits<Type>>, Utf8SliceBase {
  using Base = SliceBase<Type, SliceCodeunits<Type>>;
  using offset_type = typename Base::offset_type;
  using Base::Base;

  bool Transform(const uint8_t* input, offset_type input_string_ncodeunits,
                 uint8_t* output, offset_type* output_written) {
    int64_t output_written_64;
    bool res = Utf8SliceBase::Transform(input, input_string_ncodeunits, output, &output_written_64);
    *output_written = static_cast<offset_type>(output_written_64);
    return res;
  }
```

Reply:
Yes, good catch. Put it in a function now.